Warn when Unit object is discarded

abgruszecki opened the pull request:
Currently this kind of code is possible to write:
```scala
def either(b: Boolean): Either[Unit, Unit] = if (b) Right(Unit) else Left(Unit)
```
This is not only terrible practice, it is also misleading for someone who does not know how to create a value of type `Unit`. After the PR, a warning would be emitted:
```
warning: discarded companion object Unit of type Unit.type. Use () to obtain a value of type Unit.
       def either(b: Boolean): Either[Unit, Unit] = if (b) Right(Unit) else Left(Unit)
                                                                 ^
```
The warning intentionally takes precedence over the `-Ywarn-value-discard` one. Currently with `-Ywarn-value-discard`, a somewhat-misleading warning about a non-unit value being discarded is emitted, so it seems like the warning added in this PR is preferrable.

See discussion & longer explanations here: https://contributors.scala-lang.org/t/warn-when-unit-is-used-where-is-expected/1663 .

Note: to be honest, I have no idea what does `isThisTypeResult` represent. It seems to me that if the normal discard value warning shouldn't be emitted if it is `true`, this one should not be emitted either, but I may be wrong.

Commit:
Warn when Unit object is discarded

Precaution against accidentally using `Unit` object as a value of type
`Unit`.

hrhino commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1091,12 +1091,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             def isThisTypeResult = (tree, tree.tpe) match {
               case (Apply(Select(receiver, _), _), SingleType(_, sym)) => sym == receiver.symbol
               case _ => false
             }
```
`isThisTypeResult` is meant to detect situations like 

```scala
trait Foo { def meh: this.type }

def test(foo: Foo): Unit = foo.meh
```

and not warn. The assumption is that `this.type` doesn't give you any new information you didn't already have, so it's okay to discard. JSYK.

abgruszecki replied:
I see. In that case, it's pointless for the new warning. I'll adjust the code.

Commit:
Skip useless check

`isThisTypeResult` is meant to prevent -Ywarn-value-discard warnings
in cases such as:

```
trait Foo { def meh: this.type }

def test(foo: Foo): Unit = foo.meh
```

Since this cannot currently occur for the discarded Unit object warning,
we skip the check.

hrhino commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1096,12 +1096,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-              else if (settings.warnValueDiscard)
-                context.warning(tree.pos, "discarded non-Unit value")
-            }
+            if (tree.tpe.termSymbol == UnitClass.companionModule)
```
You could probably add `UnitMod` and friends to `definitions` if you wanted.

abgruszecki replied:
Will do!

hrhino commented:
/sync

hrhino commented:
/sync

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             }
-            if (!isThisTypeResult) context.warning(tree.pos, "discarded non-Unit value")
+            if (tree.tpe.termSymbol == UnitClass.companionModule)
+              context.warning(tree.pos, "discarded companion object Unit of type Unit.type. Use () to obtain a value of type Unit.")
```
Or pedantically, `Use '()' for the value of type Unit.` There is only one value, and I could imagine misunderstanding the current text to mean I should write `Unit()` to produce a value. The word `type` might sound ambiguous, that it's the companion of type Unit.type. `discarded Unit, the companion object of class Unit. Use '()' for the unit value, the only value of type Unit.` or `the only value of class Unit`. WDYT? Nice to see this go from forum to PR.

abgruszecki replied:
Yeah, removing the "of type Unit.type" part definitely makes sense. After looking at both wordings, I think the intent would be clearer if we just avoided the "use" word:
```
discarded Unit, the companion object of class Unit. The only value of type Unit is `()`
```
Do you think it's better that way?

som-snytt replied:
Error messages are definitely a harder problem than naming in comp sci. I prefer when it tells me what to do. There is a condescending message about "Did you mean to write...?" but I don't recommend it. `The unit value is written '()'.` Or `Write '()' for the unit value.` `unit value` is the name for it; I guess it's excessive to explain what it means.

Commit:
Add UnitModule to Definitions.scala

Commit:
Modify text of discarded Unit object warning

AleksanderBG commented:
I added a `UnitModule` to definitions - this required a bit more gymnastics than I expected, and I'm not entirely sure I got it correct.

I also changed the warning text to a middle ground between the warnings we discussed - I think it allows people to understand just fine what needs to be changed so that the warning disappears, but it doesn't really let them understand _why_ the code even compiles in the first place. This is a very good candidate for the extended warnings proposed on Contributors - in particular, from my experience corridor testing this warning on people it's crucial to explain that `Some(Unit) : Option[Unit]` is rewritten to `Some({ Unit; () })` due to the exact same mechanism which "drops" values in Unit-returning methods. We don't really have enough space here to do that.

AleksanderBG commented:
@som-snytt could I get a re-review on this? In case it wasn't clear from my babble above, I think the warning as committed right now is OK and I can adjust it if necessary.

The `UnitModule` change I'm less and less sure of - does it actually maintain binary compatibility with `2.12` if `Definitions` are exposed in reflect API?

Also it'd be nice if we could re-label this PR for `2.12.7` instead of `2.13`.

som-snytt reviewed (approved):
It's not symmetric that valueClassCompanion has different access from valueClassSymbol, but oh well.

Maybe for "Write ()" we should say "Inscribe" or "Type the unicode characters". Just kidding.

I see for "discarded non-Unit value" we could say, "discarded non-unit-value." Can't decide if the second hyphen is warranted. But also kind of kidding.

I thought maybe hrhino would want to jump back in, but otherwise LGTM!

som-snytt commented:
I see your point about the reflect API, but it's not API, it's `internal` and it's practically deprecated. I suppose they need new API for the next Coursera course and corporate training offerings. Also sort of kidding.

hrhino commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -138,6 +145,8 @@ trait Definitions extends api.StandardDefinitions {
     lazy val DoubleTpe    = DoubleClass.tpe
     lazy val BooleanTpe   = BooleanClass.tpe
 
+    lazy val UnitModule   = valueClassCompanion(tpnme.Unit.toTermName)
```
`nme.Unit` would be my only nit. (There's pun potential there, I'm sure.)

som-snytt replied:
Yes, I confess to the pun, Unit == "You nit!"

hrhino reviewed (changes requested):
Sorry, I couldn't jump back in right after eating Easter brunch. It gives you cramps, I'm told.

W/R/T binary compatibility, we've got a tool called [MiMa](https://github.com/lightbend/migration-manager) which would have failed the build if the change was unacceptably incompatible. Moving things around in `reflect.internal` annoys no one but compiler plugin writers (and maybe the next person to do a large forward merge, if they're unlucky).

I wonder if the wording would be clearer if we stopped valuing `Unit` so highly and changed the other warning to say, "discarded value of type ... (inscribe Shift-9 Shift-0 to suppress)", and then didn't warn when type was `Unit`. Then this would be "discarded value of type `Unit.type`", so maybe not.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           @inline def tpdPos(transformed: Tree) = typedPos(tree.pos, mode, pt)(transformed)
           @inline def tpd(transformed: Tree)    = typed(transformed, mode, pt)
 
-          @inline def warnValueDiscard(): Unit = if (!isPastTyper && settings.warnValueDiscard) {
```
I would prefer to keep this new warning under the flag as well.

som-snytt replied:
I think this was intended to have the moral force of deprecating `Unit` value. I don't think we considered actually issuing it as a deprecation, which as someone recently discussed, really does just mean, Don't do that. (I think one of the participants mentioned you can't just slap at deprecated on Unit itself.)

Edit: because using `Unit` for `()` is such a newbie dumb mistake. Have pity, for pity's sake.

adriaanm replied:
Brilliant! Why not give `Unit` (the companion) the actual force of `@deprecate` to tersely point out it's compiler internal, and maybe you meant `()`? Sounds like a win-win to me. (There's plenty more of internal compiler stuff that should not be used from user-level, actually.)

abgruszecki replied:
`Unit` object can be legitimately used as a parameter to `@specialized` annotation, so I don't think it can just be deprecated. We could of course also adjust the annotation, but introducing a special case there just to avoid a special case here doesn't seem like a clear win to me.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             }
-            if (!isThisTypeResult) context.warning(tree.pos, "discarded non-Unit value")
+            if (tree.tpe.termSymbol == UnitModule)
+              context.warning(tree.pos, "discarded companion object of Unit. Write '()' for the unit value, the only value of type Unit.")
```
I'm wondering whether we need to detect this special case. Maybe it's enough to improve the warning message? Maybe an addendum like s"To avoid this warning, change the expected type to ${tree.tpe} or explicitly return the Unit value, which is written as `()`.".

(PS: I'd rather not add more lazy vals to Definitions. You could use `tree.symbol == UnitClass.companionModule` to avoid introducing `UnitModule` in definitions, maybe with a name-based guard if we're worried about performance.)

som-snytt replied:
I can understand backing out the Definitions addition; and I don't think performance suffers, because no right-minded person discards their valuable values.

I would suggest avoiding long and wrappy messages, especially in advance of backported improved messages from dotty. The long and wrappy code that leads to your door, etc.

The improved messaging system could add: `Since you are obviously a rank amateur, or perhaps a hobbyist, you may be unaware that...` etc.

adriaanm reviewed (changes requested):
Agreed this warning could be improved, but I would prefer to do so more generally.

AleksanderBG commented:
@adriaanm Did you see the discussion at https://contributors.scala-lang.org/t/warn-when-unit-is-used-where-is-expected/1663? The point here is not to improve `warn-value-discard` warning, but to tell people who are very likely mistaken about how to create a `Unit` value that they should use `()` instead of `Unit`. Keeping the new warning under a flag is self-defeating. I agree that the case here is very specific, but at the same time I've seen people hit this issue multiple times. If you're worried about this breaking someone's CI - chances are they want to get rid of this code anyway, and it's trivial to do so.

Regarding `UnitModule` - what you're suggesting is the way I initially wrote it, and I changed it due to @hrhino's suggestion. Should I revert to the initial version?

som-snytt commented:
I added the welcome label just to recognize that Adriaan hazes all the contributors with less than five years' experience by asking them to do the exact opposite of what they attempted to contribute. Really, since Paul departed, the best you can hope for is that Seth will notice your PR on some boozy evening and give you a quick merge. I realize it's no longer April Fools, but for some reason I can't stop talking like this. Seriously, this is to say the long tail of a PR is worth it for someone downstream who will benefit. Granted, we're talking about someone who used `Unit` for `()`.

Has anyone asked yet why there is no `@inline def unit = ()` in `Predef`?

adriaanm commented:
@som-snytt, I didn't realize you started talking like this just recently.

adriaanm commented:
@AleksanderBG, sorry if my comments were too negative. I really your making Scala more user-friendly -- this is the area where we need help the most urgently, and that stands to make the biggest positive impact on Scala adoption!

How do you feel about going the `@deprecate` route for this sneaky companion?

AleksanderBG commented:
@adriaanm no worries - I'm keeping tabs on the ratio of LoC changed to the discussion we've had! If we remove `UnitModule` and add scala-contributors to the mix, we're bound to hit some kind of record :smile: 

Regarding `@deprecation` route - see my comment above - I'd be all for it if it could be cleanly done, but that doesn't seem to be the case.

adriaanm commented:
I'm more than happy with a thorough discussion that ends up changing the right lines :-) 

That said, I didn't find your argument against deprecating the Unit companion object. The object is useless as far as users are concerned. The box/unbox stuff won't trigger warnings, as they are only called from synthetic code. I agree equating `()` and `Unit` as values is too tricky, and not worth it anyway. `()` should actually equal `Tuple0` (hence the empty parens notation).

AleksanderBG commented:
Should've just linked to it: https://github.com/scala/scala/pull/6465#discussion_r178927633

Here's the text:
> Unit object can be legitimately used as a parameter to @specialized annotation, so I don't think it can just be deprecated. We could of course also adjust the annotation, but introducing a special case there just to avoid a special case here doesn't seem like a clear win to me.

som-snytt commented:
`Discarding Unit.type. Did you mean new Tuple0()?`

hrhino commented:
It can go along with the (new in 2.13)

```
final class Neither[A, B] private
```

adriaanm commented:
Good point about `@specialized` -- that's kind of a pity... If we're going to special case this, maybe I would prefer something along the lines that warns when `Unit`-the-companion is used as a run-time value (which would still allow use as an annotation arg).

som-snytt commented:
I spent too much time trying out slapping `compileTimeOnly` on `Unit` instead of a deprecation. Maybe it's close.

The relevant bits are in `RefChecks`, don't do `compileTimeOnly` check for annotations, which possibly makes some sense, unless that breaks the use case?

Since `Unit` is `CTO`, its usages in support of `serializable` are also `CTO`-guarded. But `specialized` is used by the compiler itself in `Definitions`, so that is special-cased in `RefChecks`.

I did the fancy bootstrap two-step, but ran out of steam and time, so maybe there are typos in the PR.

The error message is in the annotation on `Unit`. See the test.

AleksanderBG commented:
Thanks, @som-snytt! Do we close this PR in favour of yours?

som-snytt commented:
@AleksanderBG I don't know. There are hokey or hoaxy aspects to my attempt. Adriaan might be sufficiently cowed by it.

In any case, the error message consultant hired by scala center will want to review the transcript of this PR.

adriaanm commented:
Let's rally behind #6490. I'll find someone to help a bit with the mechanics.
